pkg/engine/engine.go

```diff
@@ -392,6 +458,17 @@ func Start(ctx context.Context, options ...EngineOption) (*Engine, error) {
 // more sources may be scanned by the engine.
 func (e *Engine) Finish(ctx context.Context) error {
 	defer common.RecoverWithExit(ctx)
+
+	// Handle the fail-fast termination signal.
+	e.termSignal.mutex.Lock()
+	isFailFast := e.termSignal.isFailFast
```

Comment:
is there a reason why `isFailFast` is stored in a variable first instead of checking it directly in the conditional? and is there some chance that ` e.termSignal.isFailFast` could change between when it's unlocked and when we check `isFailFast`?

Reply:
Correct. isFailFast can be modified within `notifyResults` which is run by multiple workers. This way we are ensuring the state hasn't been changed.

On a slightly related note, if we remove the concurrent notification, we could probably do away with the mutex altogether.

Reply:
Although, actually we would still need the mutex in order to use the condition variable :(